pyro/poutine/trace_messenger.py

```diff
@@ -118,7 +118,7 @@ def _reset(self):
         super()._reset()
 
     def _pyro_post_sample(self, msg):
-        if not self.param_only:
+        if not self.param_only and not msg["infer"].get("_do_not_trace"):
```

Comment:
This is a bit of a hack, but the simplest alternative I can think of is to add lots of logic to `get_importance_trace()` and MCMC utilities to discard relevant traced values. I am open to alternatives that cost < 5loc.

Reply:
This seems fine but can you add an assertion
```
assert msg["infer"].get("is_auxiliary") and not msg["is_observed"])
```
and set `is_auxiliary` in `ConjugateReparam.__call__`?

Reply:
done.